src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/Navigation.java

```diff
@@ -229,60 +186,58 @@ private void addMainNavLinks(Content target) {
-                    addPackageSummaryLink(target);
-                    addPageLabel(target, contents.classLabel, true);
-                } else {
-                    addPackageOfElementLink(target);
-                    addItemToList(target, navLinkClass);
+                // Class-use page is still generated for deprecated classes with
+                // -nodeprecated option, make sure not to link to non-existent page.
+                if (!options.noDeprecated() || !configuration.utils.isDeprecated(element)) {
+                    addPageElementLink(target);
                 }
                 addActivePageLink(target, contents.useLabel, options.classUse());
-                if (element instanceof PackageElement) {
-                    addItemToList(target, links.createLink(DocPaths.PACKAGE_TREE, contents.treeLabel));
-                } else {
-                    addItemToList(target, configuration.utils.isEnclosingPackageIncluded((TypeElement) element)
-                            ? links.createLink(DocPath.parent.resolve(DocPaths.PACKAGE_TREE), contents.treeLabel)
-                            : links.createLink(pathToRoot.resolve(DocPaths.OVERVIEW_TREE), contents.treeLabel));
+                if (options.createTree()) {
+                    if (element instanceof PackageElement) {
+                        addItemToList(target, links.createLink(DocPaths.PACKAGE_TREE, contents.treeLabel));
+                    } else {
+                        addItemToList(target, configuration.utils.isEnclosingPackageIncluded((TypeElement) element)
+                                ? links.createLink(DocPath.parent.resolve(DocPaths.PACKAGE_TREE), contents.treeLabel)
+                                : links.createLink(pathToRoot.resolve(DocPaths.OVERVIEW_TREE), contents.treeLabel));
+                    }
                 }
                 addPreviewLink(target);
                 addNewLink(target);
                 addDeprecatedLink(target);
                 addIndexLink(target);
+                addSearchLink(target);
                 addHelpLink(target);
                 break;
             case TREE:
                 addOverviewLink(target);
-                if (element == null) {
-                    addPageLabel(target, contents.moduleLabel, configuration.showModules);
-                    addPageLabel(target, contents.packageLabel, true);
-                } else {
-                    addModuleOfElementLink(target);
-                    addPackageSummaryLink(target);
+                if (element != null && !(element instanceof ModuleElement)) {
+                    addPageElementLink(target);
+                    if (options.classUse()) {
+                        if (element instanceof PackageElement || element instanceof TypeElement) {
+                            addItemToList(target, links.createLink(DocPaths.PACKAGE_USE, contents.useLabel));
+                        }
+                    }
```

Comment:
As a matter of policy, it is generally incorrect to use `instanceof` for subtypes of `Element`.  The correct coding style is to get and check the `ElementKind`.

That being said, the code here *will* work on the JDK implementation of the Language Model API, but the coding style may not work on other implementations.